lib/df-detection.c

```diff
@@ -0,0 +1,275 @@
+		 void *kernel_addr)
+{
+	if (!in_task() || current->addresses == NULL || current->pairs == NULL ||
+	    addr > TASK_SIZE)
+		return;
+	char buf[64], buf_caller[64];
+	int size;
+	size = scnprintf(buf, sizeof(buf), "%ps", caller);
+	if (len > MAX_LEN || addr == 0 || strnstr(buf, "copy_from_user_nmi", size))
+		return;
+	if (current->num_read >= current->sz && current->sz < DF_MAX_RECORDS) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		current->addresses[current->num_read].stack =
+		    df_save_stack(GFP_NOWAIT);
+		detect_intersection(kernel_addr);
+		current->num_read++;
+	}
+}
+
+void start_system_call(long syscall)
+{
+	if (!current->df_enable)
+		return;
+	current->syscall_num = syscall;
+	current->addresses = (struct df_address_range *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
+	current->sz = current->addresses ? DF_INIT_SIZE : 0;
+	current->num_read = 0;
+	current->pairs = (struct df_pair *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
+	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+	current->df_index = 0;
+}
+
+void end_system_call(void)
+{
+	if (!current->df_enable)
+		return;
+	if (current->pairs != NULL) {
+		if (current->df_index)
+			report();
+		kfree(current->pairs);
+		current->pairs = NULL;
+		current->df_index = 0;
+		current->df_size = 0;
+	}
+	if (current->addresses != NULL) {
+		current->num_read = 0;
+		current->sz = 0;
+		kfree(current->addresses);
+		current->addresses = NULL;
+	}
+}
+
+void report(void)
+{
+	if (!check_valid_detection())
+		return;
+	int i;
+	unsigned long *first_entries;
+	unsigned int first_nr_entries;
+	unsigned long *second_entries;
+	unsigned int second_nr_entries;
+	unsigned long first_frame = 0;
+	unsigned long second_frame = 0;
+
+	for (i = 0; i < current->df_index; i++) {
+		if (current->addresses[current->pairs[i].first].stack &&
+		    current->addresses[current->pairs[i].second].stack) {
+			first_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].first].stack,
+			    &first_entries);
+			second_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].second].stack,
+			    &second_entries);
+			int first_index =
+			    filter_stack(first_entries, first_nr_entries);
+			int second_index =
+			    filter_stack(second_entries, second_nr_entries);
+			first_frame = first_entries[first_index];
+			second_frame = second_entries[second_index];
+			char first_buf[64], second_buf[64];
```

Comment:
Linux kernel style doesn't allow mixing declarations and code. You can either move these two to the top of the scope, or to the declaration list at the beginning of the function.